---
title: "G3viz: Interactively visualize genomic mutation data using a lollipop-diagram"
date: "2019-03-01"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{G3viz: interactively visualize genomic mutation data using a lollipop-diagram}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{utf8}
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
	echo = TRUE,
	collapse = TRUE,
	comment = "#>"
)
```

Easily and effectively visualizing genomic mutation data can help researchers to better understand the data and validate the findings.
`G3viz` is a R-package which provides an easy-to-use lollipop-diagram tool.  It enables users to interactively visualize annotated genomic mutation data in Rstudio or a web browser, without having to know any HTML5/JavaScript technologies.  The features of `g3viz3` include

- easily retrieve protein domain information for the given gene
- built-in function to map from mutation annotation types (a.k.a, variant classification; _e.g._, Frame\_Shift\_Del, In\_Frame\_Ins) to mutation annotation classes (_e.g._, Inframe/Missense/Truncating)
- interactive chart legend
- zoom and brush tools to view every details
- highlighting findings in charts
- rich chart options, with easy-to-use built-in chart themes 
- over [30 color schemes](https://bl.ocks.org/phoeguo/81dffe0c7c6c8caae06f6a5f60c70d19)
- save charts to publication-quality figures in SVG or PNG format

## Table of contents
1. [Install package](#install)

1. [Quick Start](#quickstart)
    * [Example 1](#ex1): visualize genomic mutation data from `MAF` file
    * [Example 2](#ex2): visualize genomic mutation data from `CSV` or `TSV` file
    * [Example 3](#ex3): visualize genomic mutation data from `cBioPortal`

1. [Usage](#usage)
    * [Input data format](#inputdata)
    * [Pfam domain information](#pfamdomain)
    * [Read data](#readdata)
    * [Color scheme](#colorscheme)
    * [Set chart options](#options)
    * [Generate g3-lollipop diagram](#view)
    * [Save charts and html](#output)
    
1. [Sessioninfo](#sessioninfo)


## 1. <a name="install"></a>Install package
Install package from R repository
```r
# install package
install.packages("g3viz", repos = "http://cran.us.r-project.org")
```
or if you prefer to install development version from github
```r
# Install devtools
if("devtools" %in% rownames(installed.packages()) == FALSE){ 
  install.packages("devtools")
}

# install from github
devtools::install_github("g3viz/g3viz")
```

```{r include = FALSE}
# install package
library(g3viz)
```


## 2. <a name="quickstart"></a>Quick Start

### <a name="ex1"></a>Example 1: Visualize genomic mutation data from `MAF` file

Mutation Annotation Format ([MAF](https://docs.gdc.cancer.gov/Data/File_Formats/MAF_Format/)) is a widely-used tab-delimited text file with annotated genomic mutation information.  The translational effect of variant alleles in `MAF` files are usually in the column of `Variant_Classification` (i.e., `Frame_Shift_Del`, `Split_Site`).
The input somatic mutation data of this example was downloaded directly from GDC Data Portal for the [TCGA-BRCA](https://portal.gdc.cancer.gov/projects/TCGA-BRCA) project.
The data was visualized according to the parsed `Mutation_Class` information.

```{r, include = TRUE}
# load MAF file
maf.file <- system.file("extdata", "TCGA.BRCA.varscan.somatic.maf.gz", package = "g3viz")

# read in MAF file
# Note: "Mutation_Class" information is parsed according to "Variant_Classification" column
mutation.dat <- readMAF(maf.file)

# chart 1: using default chart options (except chart.width)
chart1.options <- g3Lollipop.options(chart.width = 680)

g3Lollipop(mutation.dat,
           gene.symbol = "PIK3CA",
           plot.options = chart1.options,
           factor.col = "Mutation_Class")

# chart 2: add some chart options, and visualize according to "Variant_Classification" information
chart2.options <- g3Lollipop.options(# chart options
                                    chart.width = 680,
                                    chart.type = "pie",
                                    chart.margin = list(left = 40, right = 20, top = 30, bottom = 25),
                                    chart.background = "#deebf7",
                                    # lollipop axis
                                    y.axis.label = "# of mutations",
                                    axis.label.font = "bold 16px Sans",
                                    axis.label.alignment = "middle",
                                    axis.label.dy = "-1em",
                                    # legend options
                                    legend.title = "Variant classification",
                                    # lollipop track options
                                    lollipop.track.height = 300,
                                    lollipop.track.background = "white",
                                    lollipop.pop.info.color = "white",
                                    lollipop.color.scheme = "set3",
                                    # domain annotation track options
                                    domain.color.scheme = "google5",
                                    domain.text.color = "white",
                                    # title
                                    title.text = "Tumor protein p53 (TP53, from TCGA-BRCA project)", 
                                    title.font = "normal 16px Verdana",
                                    title.color = "steelblue", 
                                    title.alignment = "middle",
                                    title.dy = "0.3em",
                                    # domain brush options
                                    brush.selection.background = "steelblue",
                                    brush.selection.opacity = 0.3,
                                    brush.border.color = "#666666",
                                    brush.handler.color = "#666666"
                                    )

g3Lollipop(mutation.dat,
           gene.symbol = "TP53",
           factor.col = "Variant_Classification",
           # style of output buttons
           btn.style = "blueBtn",
           plot.options = chart2.options)
```

[↥ back to top](#top)


### <a name="ex2"></a>Example 2: visualize genomic mutation data from `CSV` or `TSV` file

Loading annotated genomic mutation data from `CSV` or `TSV` format, this example mimics the lollipop diagram as of [cBioPortal MutationMapper](http://www.cbioportal.org/mutation_mapper).


```{r, include=TRUE}
# load data
mutation.csv <- system.file("extdata", "ccle.csv", package = "g3viz")

# read in data
# specify "gene.symbol.col", "variant.class.col", and "protein.change.col"
mutation.dat <- readMAF(mutation.csv,
                        gene.symbol.col = "Hugo_Symbol",
                        variant.class.col = "Variant_Classification",
                        protein.change.col = "amino_acid_change",
                        sep = ",")  # separator of csv file

# try to mimic MutationMapper (http://www.cbioportal.org/mutation_mapper.jsp)
plot.options <- g3Lollipop.options(
                                   # chart options
                                   chart.width = 680,
                                   chart.type = "circle",
                                   # lollipop track options
                                   lollipop.track.height = 200,
                                   lollipop.track.background = "transparent",
                                   lollipop.pop.max.size = 4,
                                   lollipop.pop.min.size = 4,
                                   # set larger than lollipop.pop.max.size to turn off pop info
                                   lollipop.pop.info.limit = 4.1, 
                                   # y-axis label
                                   y.axis.label = "# of TP53 Mutations",
                                   lollipop.line.color = "grey",
                                   lollipop.line.width = 0.5,
                                   lollipop.circle.color = "grey",
                                   lollipop.circle.width = 0.5,
                                   lollipop.color.scheme = "bottlerocket2",
                                   # domain annotation bar
                                   anno.bar.fill = "#969696",
                                   anno.bar.margin = list(top = 4, bottom = 8),
                                   # domain track options
                                   domain.margin = list(top = 2, bottom = 6),
                                   domain.color.scheme = "darjeeling2",
                                   domain.text.font = "normal 12px Arial",
                                   domain.text.color = "white",
                                   # disable brush
                                   brush = FALSE,
                                   # disable legend
                                   legend = FALSE)

g3Lollipop(mutation.dat,
           gene.symbol = "TP53",
           protein.change.col = "amino_acid_change",
           save.png.btn = FALSE,
           save.svg.btn = FALSE,
           plot.options = plot.options)
```


[↥ back to top](#top)


### <a name="ex3"></a>Example 3: visualize genomic mutation data from `cBioPortal`
[cBioPortal](http://www.cbioportal.org/) provides download for many cancer genomics data sets.
The package provides an easy way to retrieve from this portal.  For example, to retrieve genomic mutation data of the [msk\_impact\_2017](https://www.ncbi.nlm.nih.gov/pubmed/28481359) study for the gene `BRCA1` and visualize the data using `g3viz`

```{r, include=TRUE}
# get genomic mutation data of "msk_impact_2017"" study from cBioPortal
mutation.dat <- g3viz::getMutationsFromCbioportal("msk_impact_2017", "BRCA1")

# lollipop diagram with default options
g3Lollipop(mutation.dat, gene.symbol = "BRCA1")
```

#### Note:
- Internet access is required to download data from [cBioPortal](http://www.cbioportal.org/), and it may take more than 10 seconds.
- To list all studies of cBioPortal
```r
cgds <- cgdsr::CGDS("http://www.cbioportal.org/public-portal/")
all.studies <- cgdsr::getCancerStudies(cgds)

# Pick a "caner_study_id" (contain a mutation data set)
# then, query the data using a primary HGNC gene symbol
mutation.dat <- g3viz::getMutationsFromCbioportal("all_stjude_2016", "TP53")
```


[↥ back to top](#top)










## 3. <a name="usage"></a>Usage

1.  <a name="readdata"></a>Read data

Genomic mutation data (_e.g._, [aggregated somatic mutations](https://docs.gdc.cancer.gov/Encyclopedia/pages/Aggregated_Somatic_Mutation/)) can be loaded in three ways

  * from [MAF](https://docs.gdc.cancer.gov/Data/File_Formats/MAF_Format/) file, for example,

```r
maf.file <- system.file("extdata", "TCGA.BRCA.varscan.somatic.maf.gz", package = "g3viz")
mutation.dat <- readMAF(maf.file)
```

  * directly from [cBioPortal](http://www.cbioportal.org/) (internet access required), for example,

```r
# get mutation data of msk_impact_2017 study from cBioPortal
mutation.dat <- getMutationsFromCbioportal("msk_impact_2017", "TP53")
```

  * from `CSV` or `TSV` formatted file

```r
# load and read data
mutation.csv <- system.file("extdata", "ccle.csv", package = "g3viz")

mutation.dat <- readMAF(mutation.csv,
                        gene.symbol.col = "Hugo_Symbol",
                        variant.class.col = "Variant_Classification",
                        protein.change.col = "amino_acid_change",
                        sep = ",")  # separator of csv file
```


[↥ back to top](#top)


3.  <a name="colorscheme"></a>Ordinal color schemes

The `g3viz` package is built in with over 30 color schemes as shown in [this example](https://bl.ocks.org/phoeguo/81dffe0c7c6c8caae06f6a5f60c70d19).
Examples of applying these color schemes to lollipop pops and domain composition are shown in [demo 1](https://bl.ocks.org/phoeguo/raw/2868503a074a6441b5ae6d987f150d48/) and 
[demo 2](https://bl.ocks.org/phoeguo/raw/de79b9ce9bda958173af9891ab7aec93/), respectively.



2.  <a name="options"></a>Set chart options

Chart options can be specified using `g3Lollipop.options()` function (_e.g._, `g3Lollipop.options(chart.type = "circle", lollipop.track.background = "transparent")`.  Use `?g3viz::g3Lollipop.options` to check these options.
These options are listed in the following table. 

Option name             | Description
----------------------- | --------------------------------------
chart.width             | chart width in px.  Default `800`.  Note: `chart.height` is calculated by adding `lollipop.track.height` and `anno.height`.
chart.type              | pop type, _pie_ or _circle_.  Default `pie`.
chart.margin            | specify chart margin in _list_ format.  Default `list(left = 40, right = 20, top = 15, bottom = 25)`.
chart.background        | chart background.  Default `transparent`.
transition.time         | chart animation transition time in millisecond.  Default `600`.
y.axis.label            | Y-axis label text.  Default `mutations`.
axis.label.font         | css font style shorthand (font-style font-variant font-weight font-size/line-height font-family).  Default `normal 12px Arial`.
axis.label.color        | axis label text color.  Default `#4f4f4f`.
axis.label.alignment    | axis label text alignment (start/end/middle). Default `middle`
axis.label.dy           | text adjustment of axis label text.  Default `-2em`.
legend.margin           | legend margin in _list_ format. Default `list(left = 10, right = 0, top = 5, bottom = 5)`.
legend.interactive      | legend interactive mode. Default `TRUE`.
legend.title            | legend title. If `NA`, use factor name as `factor.col`. Default is `NA`.
lollipop.track.height   | height of lollipop track. Default `420`.
lollipop.track.background  | background of lollipop track. Default `rgb(244,244,244)`.
lollipop.pop.min.size  | lollipop pop minimal size in px. Default `2`.
lollipop.pop.max.size | lollipop pop maximal size in px. Default `12`.
lollipop.pop.info.limit | threshold of lollipop pop size to show count information in middle of pop. Default `8`.
lollipop.pop.info.color | lollipop pop information text color. Default `#EEE`.
lollipop.line.color | lollipop line color. Default `rgb(42,42,42)`.
lollipop.line.width | lollipop line width. Default `0.5`.
lollipop.circle.color | lollipop circle border color. Default `wheat`.
lollipop.circle.width | lollipop circle border width. Default `0.5`.
lollipop.label.ratio | lollipop click-out label font size to circle size ratio. Default `1.4`.
lollipop.label.min.font.size | lollipop click-out label minimal font size. Default `10`.
lollipop.color.scheme | color scheme to fill lollipop pops. Default `accent`. [demo1](https://bl.ocks.org/phoeguo/raw/2868503a074a6441b5ae6d987f150d48/) or [demo2](https://bl.ocks.org/phoeguo/raw/de79b9ce9bda958173af9891ab7aec93/).
title.text | title of chart. Default "".
title.font | font of chart title. Default `normal 16px Arial`.
title.color | color of chart title. Default `#424242`.
title.alignment | text alignment of chart title (start/middle/end). Default `middle`.
title.dy | text adjustment of chart title. Default `0.35em`.
anno.height | height of protein structure annotation track. Default `30`.
anno.margin | margin of protein structure annotation track. Default `list(top = 4, bottom = 0)`.
anno.background | background of protein structure annotation track. Default `transparent`.
anno.bar.fill | background of protein bar in protein structure annotation track. Default `#E5E3E1`.
anno.bar.margin | margin of protein bar in protein structure annotation track. Default `list(top = 2, bottom = 2)`.
domain.color.scheme | color scheme of protein domains. Default `category10`.
domain.margin | margin of protein domains. Default `list(top = 0, bottom = 0)`.
domain.text.font | domain label text font in shorthand format. Default `normal 11px Arial`.
domain.text.color | domain label text color. Default `#F2F2F2`.
legend | if show legend. Default `TRUE`.
tooltip | if show tooltip. Default `TRUE`.
brush | if show brush. Default `TRUE`.
zoom | if enable zoom feature. Default `TRUE`.


[↥ back to top](#top)


3.  <a name="view"></a>Visualize genomic mutation data via a lollipop diagram in a web browser

Call `g3Lollipop` function to visualize genomic mutation data in a web browser.  For example

```r
g3Lollipop(mutation.dat,
           gene.symbol = "APC",
           gene.symbol.col = gene.symbol.colname,
           protein.change.col = protein.change.colname,
           plot.options = plot.options)
```


[↥ back to top](#top)

## 4. <a name="sessioninfo"></a>Session Info

```{r, include=TRUE}
sessionInfo()
```

[↥ back to top](#top)
